query-engine/prisma/src/request_handlers/graphql/handler.rs

```diff
@@ -88,13 +112,90 @@ async fn handle_single_query(query: SingleQuery, ctx: Arc<PrismaContext>) -> Pri
     PrismaResponse::Single(responses)
 }
 
-async fn handle_graphql_query(body: SingleQuery, ctx: &PrismaContext) -> PrismaResult<response_ir::Responses> {
-    debug!("Incoming GQL query: {:?}", &body.query);
-    debug!("Operation: {:?}", body.operation_name);
+async fn handle_batch(queries: Vec<Operation>, ctx: &Arc<PrismaContext>) -> PrismaResponse {
+    let mut futures = Vec::with_capacity(queries.len());
 
-    let gql_doc = gql::parse_query(&body.query)?;
-    let query_doc = GraphQLProtocolAdapter::convert(gql_doc, body.operation_name)?;
+    for operation in queries.into_iter() {
+        futures.push(tokio::spawn(handle_single_query(operation, ctx.clone())));
+    }
+
+    let responses = future::join_all(futures)
+        .await
+        .into_iter()
+        .map(|res| res.expect("IO Error in tokio::spawn"))
+        .collect();
+
+    PrismaResponse::Multi(responses)
+}
+
+async fn handle_compacted(document: CompactedDocument, ctx: &Arc<PrismaContext>) -> PrismaResponse {
+    use user_facing_errors::Error;
+
+    let plural_name = document.plural_name();
+    let singular_name = document.single_name();
+    let keys = document.keys;
+    let arguments = document.arguments;
+    let nested_selection = document.nested_selection;
+
+    match AssertUnwindSafe(handle_graphql_query(document.operation, ctx))
+        .catch_unwind()
+        .await
+    {
+        Ok(Ok(mut responses)) => {
+            // We find the response data and make a hash from the given unique
+            // keys.
+            let mut data = responses
+                .take_data(plural_name)
+                .unwrap()
+                .into_list()
+                .unwrap()
+                .index_by(keys.as_slice());
+
+            let results = arguments
+                .into_iter()
+                .map(|args| {
+                    let vals: Vec<QueryValue> = args.into_iter().map(|(_, v)| v).collect();
+                    let mut responses = Responses::with_capacity(1);
+
+                    match data.remove(&vals) {
```

Comment:
what if we have multiple findOnes with the same keys? Here we have to clone instead of `remove`.

Reply:
Or we have a `Cow` in reponse IR if possible.